csharp/src/Google.Protobuf/CodedInputReader.cs

```diff
@@ -0,0 +1,762 @@
+﻿#region Copyright notice and license
+// Protocol Buffers - Google's data interchange format
+// Copyright 2008 Google Inc.  All rights reserved.
+// https://developers.google.com/protocol-buffers/
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#endregion
+
+#if GOOGLE_PROTOBUF_SUPPORT_SPAN
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Google.Protobuf
+{
+    // TODO: class name TBD (CodedInputReader, CodedInputContext, CodedInputByRefReader, WireFormatReader)
+    public ref struct CodedInputReader
```

Comment:
Just FYI, we've recently made some optimizations to CodedInputStream in https://github.com/protocolbuffers/protobuf/pull/6843 (motivated by internal needs),
so you might wanna mirror them at some point.
That brings up the challenge that we should try to avoid code duplication where possible, because it makes the codebase much more complex and harder to test, so I'd welcome ideas for refactoring that would avoid "code duplication" between CodedInputStream and CodedInputReader (the fact that the code is not exactly the same makes it more challenging, but if feels wrong having two copies of the almost same parsing code which tends to be quite intricate and also needs to be heavily optimized).

Reply:
I've already de-duplicated as much as I know how. These are fundamentally different types (class vs ref struct). There is no common interface that they share.

There might be some minor parts that don't directly touch parsing/writing that can be refactored into shared types, but not very much.

Reply:
Are you willing to change the array version to be span based? Maybe then internal helpers could be shared between both.

Reply:
CodedInputStream/CodedOutputStream aren't ref structs so they can't hold a span. And remember that these need to work on .NET Framework. Replacing array with span there will hurt performance.

Besides, most of the internal code is different between the implementations. The new reader/writer use a lot of the new BinaryPrimative/BitConverter helper methods, and strings use the new span overloads on Encoding.